### PR TITLE
Clustering

### LINKED ISSUE #110: Review: Ch 10 (clustering)
# Reviewer E: 

- [x] The dataset here is strange because “loyalty” and “satistfaction” are already the “latent” concepts and not things that feel directly measurable
- [x] As I describe in 3 (*see here: #103*), consider using summarize_*() variants instead of map_df()
- [ ] ~~Why is the elbow method done manually instead of in the tidymodels framework? Example here: https://www.tidymodels.org/learn/statistics/k-means/#clustering-in-r . It may be good to make students work to internalize formula but if so should at least highlight other option for long term~~
- [ ] Unsupervised learning is often quite hard in practice and may not always find meaningful results. It could be useful in this section especially to show an example of the method struggling or failing

### PR DESCRIPTION
**Pull request by ttimbers**
@trevorcampbell @leem44 - this PR is the final review of the clustering chapter. It should close #110 . Can one of you please do a review of this? At least a light one?

### CODE REVIEW

File: clustering.Rmd [hunk 1]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
-be nearly impossible to annotate all the articles on Wikipedia with human-made
-topic labels. However, we can still cluster the articles without this
-information to find groupings corresponding to topics automatically. 
-
-However, because there is no predictive target, it is not as easy to evaluate
+Note that clustering is a fundamentally different kind of task 
+than classification or regression. 
+In particular, both classification and regression are *supervised tasks* 
+\index{classification}\index{regression}\index{supervised} 
+where there is a *response variables* (a category label or value),
```
Comment (trevorcampbell) [inline clustering.Rmd:56]: grammar "response variables" plural should be singular

File: clustering.Rmd [hunk 2]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+and examine the structure of data without any response variable labels 
+or values to help us. 
+This approach has both advantages and disadvantages. 
+Clustering requires no additional annotation or input on the data. 
+For example, it would be nearly impossible to annotate 
+all the articles on Wikipedia with human-made topic labels. 
+However, we can still cluster the articles without this information 
+to find groupings corresponding to topics automatically. 
+
+However, because there is no response variable, it is not as easy to evaluate
```
Comment (trevorcampbell) [inline clustering.Rmd:70]: "However" twice in close proximity awk

File: clustering.Rmd [hunk 3]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
-> guess the missing labels. This sort of task is beneficial, for example, when one 
-> has an unlabelled data set that is too large to manually label, but one is willing to
-> provide a few informative example labels as a "seed" to guess the labels for all the data.
-
+As in the case of classification, 
+there are many possible methods that we could use to cluster our observations 
+to look for subgroups. 
+In this book, we will focus on the widely used K-means \index{K-means} algorithm.
+In your future studies, you might encounter hierarchical clustering,
+principle component analysis, multidimensional scaling, and more;
```
Comment (trevorcampbell) [inline clustering.Rmd:82]: principal not principle

File: clustering.Rmd [hunk 4]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
-> has an unlabelled data set that is too large to manually label, but one is willing to
-> provide a few informative example labels as a "seed" to guess the labels for all the data.
-
+As in the case of classification, 
+there are many possible methods that we could use to cluster our observations 
+to look for subgroups. 
+In this book, we will focus on the widely used K-means \index{K-means} algorithm.
+In your future studies, you might encounter hierarchical clustering,
+principle component analysis, multidimensional scaling, and more; 
+see the additional resources section at the end of the next chapter
```
Comment (trevorcampbell) [inline clustering.Rmd:83]: end of this chapter, not the next one

File: clustering.Rmd [hunk 5]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
 
 **An illustrative example** 
 
-
 Here we will present an illustrative example using a data set \index{Palmer penguins} from the
 [{palmerpenguins} R data package](https://allisonhorst.github.io/palmerpenguins/). This data set was
 collected by [Dr. Kristen Gorman](https://www.uaf.edu/cfos/people/faculty/detail/kristen-gorman.php) and
-the [Palmer Station, Antarctica LTER](https://lternet.edu/) and includes
-measurements for adult penguins near Palmer Station [@palmerpenguins]. We have
+the [Palmer Station, Antarctica Long Term Ecological Research Site](https://pal.lternet.edu/) and includes
```
Comment (trevorcampbell) [inline clustering.Rmd:101]: this is confusing wording

"this data set was collected by [person] and [station, site] and includes msrmts..."

File: clustering.Rmd [hunk 6]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
 modified the data set for use in this chapter. Here we will focus on using two
 variables---penguin bill and flipper length, both in millimeters---to determine whether 
 there are distinct types of penguins in our data.
 Understanding this might help us with species discovery and classification in a data-driven
 way.
-```{r 09-penguins, echo = FALSE, message = FALSE, warning = FALSE, fig.height = 3, fig.width = 4, fig.cap = "Gentoo penguin. [@penguinsimage]", fig.retina = 2}
-knitr::include_graphics("img/gentoo.jpg")
+
+```{r 09-penguins, echo = FALSE, message = FALSE, warning = FALSE, fig.height = 3, fig.width = 4, fig.cap = "Gentoo penguin.", fig.retina = 2}
+knitr::include_graphics("img/Gentoo_Penguin._(8671680772).jpg")
```
Comment (trevorcampbell) [inline clustering.Rmd:110]: any chance we can keep the older / simpler filename of just `gentoo.jpg` ? parentheses in filenames is kind of annoying for common command line tasks

File: clustering.Rmd [hunk 7]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
 ```
 
-Below we will work with `penguin_data`, a subset of 18 observations of the original data, which has already been scaled. We will discuss scaling for K-means in more detail later in this chapter. \index{mutate}\index{read function!read\_csv} 
+To learn about K-means clustering in this chapter, 
+we will work with `penguin_data`, 
+a subset of 18 observations of the original data,
```
Comment (trevorcampbell) [inline clustering.Rmd:115]: incorrect comma usage in this setence

File: clustering.Rmd [hunk 8]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+we will work with `penguin_data`, 
+a subset of 18 observations of the original data, 
+which has already been standardized 
+(remember from Chapter \@ref(classification) that scaling is part of the standardization process). 
+We will discuss scaling for K-means in more detail later in this chapter. 
+\index{mutate}\index{read function!read\_csv} 
+
+Before we get started, we will load the `tidyverse` metapackage to ensure
+we have the functions we need, 
+as well as set a random seed, so that our analysis will be reproducible.
```
Comment (trevorcampbell) [inline clustering.Rmd:123]: comma misusage

File: clustering.Rmd [hunk 9]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+As we will learn in more detail later in the chapter, 
+the K-means clustering algorithm uses random numbers -
```
Comment (trevorcampbell) [inline clustering.Rmd:125]: bad hyphen usage here

File: clustering.Rmd [hunk 10]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+the K-means clustering algorithm uses random numbers - 
+which is why setting the seed here is important.
+
 ```{r 10-toy-example-data, echo = FALSE, message = FALSE, warning = FALSE}
 library(tidyverse)
 
 data <- read_csv("data/toy_penguins.csv") |>
   mutate(cluster = as_factor(cluster)) |>
-  mutate(flipper_length_scaled = scale(flipper_length_mm), bill_length_scaled = scale(bill_length_mm))
+  mutate(flipper_length_standardized = as.double(scale(flipper_length_mm)),
```
Comment (trevorcampbell) [inline clustering.Rmd:133]: `as.double` ? looks like a base R function. Is there a `as_double`? if not no big deal, just unused to seeing "dot" functions

File: clustering.Rmd [hunk 11]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+
+```{r, message = FALSE, warning = FALSE}
+penguin_data <- read_csv("data/penguins_standardized.csv")
 penguin_data
 ```
 
-```{r 10-toy-example-plot, echo = FALSE, warning = FALSE, fig.height = 4, fig.width = 4.35, fig.cap = "Subset data of penguin scaled bill length versus flipper length."}
-base <- ggplot(data, aes(x = flipper_length_scaled, y = bill_length_scaled)) +
+Next, we can create a scatter plot using this data set to see if we can use
+data visualization to detect subtypes or groups in our data set.
```
Comment (trevorcampbell) [inline clustering.Rmd:155]: awkward phrasing: "next, we can create a plot to see if we can use a plot to ..."

File: clustering.Rmd [hunk 12]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+
+```{r 10-toy-example-plot, warning = FALSE, fig.height = 4, fig.width = 4.35, fig.cap = "Subset data of penguin standardized bill length versus standardized flipper length."}
```
Comment (trevorcampbell) [inline clustering.Rmd:157]: "Subset data of penguin standardized bill length" -- this is very unclear

File: clustering.Rmd [hunk 13]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
-separate the data into groups is to use a clustering algorithm.
-In this chapter, we will focus on the *K-means* algorithm, \index{K-means} a widely-used
-and often very effective clustering method, combined with the *elbow method* \index{elbow method} for
-selecting the number of clusters. This procedure will separate the data into 
-the following groups denoted by colour:
-
-```{r 10-toy-example-clustering, echo = FALSE, warning = FALSE, fig.height = 4, fig.width = 5, fig.cap = "Subset data of penguin scaled bill length versus flipper length with coloured groups"}
-ggplot(data, aes(y = bill_length_scaled, x = flipper_length_scaled, colour = cluster)) +
+suspect there are a few subtypes of penguins 
+within the observations we have in our data set.
```
Comment (trevorcampbell) [inline clustering.Rmd:166]: simpler phrasing: "few subtypes of penguins in our data set"

File: clustering.Rmd [hunk 14]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+We can see ~ 3 groups of observations in Figure \@ref(fig:10-toy-example-plot),
```
Comment (trevorcampbell) [inline clustering.Rmd:167]: replace "~" with "roughly"  (tildes may not typeset well in PDF, and rely on students knowing tilde means roughly)

File: clustering.Rmd [hunk 15]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+including:
+
+1. a small flipper and bill length group,
+2. a small flipper length, but large bill length group, and
+3. a large  flipper and bill length group.
+
+This is interesting that we can use data visualization to uncover such patterns,
```
Comment (trevorcampbell) [inline clustering.Rmd:174]: "It is interesting"

File: clustering.Rmd [hunk 16]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+but also leads to the questions of how might we find 
+these kind of grouping patterns automatically,
```
Comment (trevorcampbell) [inline clustering.Rmd:176]: this sentence blends a statement and a question mark in an odd way... I think it's just a statement

File: clustering.Rmd [hunk 17]
```diff
@@ -45,95 +49,147 @@ use the subgroups to generate new questions about the data and follow up with a
...
+The way to rigorously separate the data into groups 
+is to use a clustering algorithm.
+In this chapter, we will focus on the *K-means* algorithm, 
+\index{K-means} a widely-used and often very effective clustering method, 
+combined with the *elbow method* \index{elbow method} 
+for selecting the number of clusters. 
+This procedure will separate the data into the following groups 
+denoted by colour:
+
+```{r 10-toy-example-clustering, echo = FALSE, warning = FALSE, fig.height = 4, fig.width = 5, fig.cap = "Subset data of penguin standardized bill length versus standardized flipper length with coloured groups"}
```
Comment (trevorcampbell) [inline clustering.Rmd:187]: same caption comment here as earlier

File: clustering.Rmd [hunk 18]
```diff
@@ -181,33 +232,38 @@ In K-means clustering, we measure the quality of a cluster by its
 First, we find the cluster centers by computing the mean of each variable 
 over data points in the cluster. For example, suppose we have a 
 cluster containing four observations, and we are using two variables, $x$ and $y$, to cluster the data.
-Then we would compute the $x$ and $y$ variables, $\mu_x$ and $\mu_y$, of the cluster center via
+Then we would compute the coordinates, $\mu_x$ and $\mu_y$, of the cluster center via:
```
Comment (trevorcampbell) [inline clustering.Rmd:235]: no need for a colon here

File: clustering.Rmd [hunk 19]
```diff
@@ -181,33 +232,38 @@ In K-means clustering, we measure the quality of a cluster by its
...
   )) +
-  geom_point(aes(y = mean(bill_length_scaled), x = mean(flipper_length_scaled), colour = "red", size = 2.5)) +
+  geom_point(aes(y = mean(bill_length_standardized), x = mean(flipper_length_standardized), colour = "red", size = 2.5)) +
   theme(legend.position = "none") 
 base
 ```
 
 The second step in computing the WSSD is to add up the squared distance \index{distance!K-means} between each point in the cluster and the cluster center.
-We use the straight-line / Euclidean distance formula that we learned about in the classification chapter.
+We use the straight-line / Euclidean distance formula that we learned about in the [classification chapter](\@ref(classification)).
```
Comment (trevorcampbell) [inline clustering.Rmd:266]: Instead of "classification chapter" keep consistent with the other refs we use elsewhere:

"that we learned about in Chapter \\@ref(classification)"

File: clustering.Rmd [hunk 20]
```diff
@@ -238,75 +294,93 @@ Note, however, that "large" is relative to *both* the scale of the variables for
 
 ### The clustering algorithm
 
-We begin the K-means \index{K-means!algorithm} algorithm by picking K, and uniformly randomly assigning data to the K clusters. 
+We begin the K-means \index{K-means!algorithm} algorithm by picking K, 
+and randomly assigning an equal number of observations
```
Comment (trevorcampbell) [inline clustering.Rmd:298]: We don't necessarily assign an equal # of obs to each cluster. We do typically uniformly assign labels tho. I would revert this change

File: clustering.Rmd [hunk 21]
```diff
@@ -238,75 +294,93 @@ Note, however, that "large" is relative to *both* the scale of the variables for
...
+Figure \@ref(fig:10-toy-kmeans-iter). 
+There each row corresponds to an iteration,
 where the left column depicts the center update, 
-and the right column depicts the reassignment of data to clusters): 
+and the right column depicts the reassignment of data to clusters.
+
 <center>
 **Center Update**  &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;**Label Update**
-```{r 10-toy-kmeans-iter, echo = FALSE, warning = FALSE, fig.height = 4, fig.width = 8}
+```{r 10-toy-kmeans-iter, echo = FALSE, warning = FALSE, fig.height = 16, fig.width = 8, fig.cap = "First four iterations of K-means clustering on the `penguin_data` example data set. Each row corresponds to an iteration,where the left column depicts the center update, and the right column depicts the reassignment of data to clusters. Cluster centers are indicated by larger points that are outlined in black."}
```
Comment (trevorcampbell) [inline clustering.Rmd:333]: missing space after comma in the caption

File: clustering.Rmd [hunk 22]
```diff
@@ -320,64 +394,92 @@ in the fourth iteration; both the centers and labels will remain the same from t
 > ways to assign the data to clusters. So at some point, the total WSSD must stop decreasing, which means none of the assignments
 > are changing, and the algorithm terminates.
 
+What kind of data is suitable for K-means clustering? 
+In the simplest version of K-means clustering that we have presented here,
+the straight-line / Euclidean distance is used to measure the
```
Comment (trevorcampbell) [inline clustering.Rmd:399]: I would just put "straight-line distance"  -- we only mention in passing that it's also called "Euclidean" when we first introduce it, but refer to it everywhere else as strgiht-line

File: clustering.Rmd [hunk 23]
```diff
@@ -320,64 +394,92 @@ in the fourth iteration; both the centers and labels will remain the same from t
...
 
+What kind of data is suitable for K-means clustering? 
+In the simplest version of K-means clustering that we have presented here,
+the straight-line / Euclidean distance is used to measure the 
+distance between observations and cluster centres. 
+This means that only quantitative data should be used with this algorithm.
+There are variants on the K-means algorithm, 
+as well as other clustering algorithms entirely, 
+that use other distance metrics 
+to allow for non-quantitative data to be clustered,
```
Comment (trevorcampbell) [inline clustering.Rmd:405]: run-on with commas

File: clustering.Rmd [hunk 24]
```diff
@@ -466,17 +570,28 @@ To address this problem, we typically standardize \index{standardization!K-means
 which ensures that each variable has a mean of 0 and standard deviation of 1.
 The `scale` function in R can be used to do this. 
 We show an example of how to use this function 
-below using an unscaled version of data set in this chapter:
+below using an unscaled, and unstandardized version of data set in this chapter.
```
Comment (trevorcampbell) [inline clustering.Rmd:573]: comma misuse

File: clustering.Rmd [hunk 25]
```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
...
-
-The solution is to create something called an *anonymous function*. \index{function!anonymous}
-An anonymous function is a function that has no name,
-unlike other functions you've seen so far (`kmeans`, `select`, etc). 
-To do this we will write our map statement like this:
-
-```{r, eval = FALSE}
-map(penguin_clust_ks, function(k) kmeans(scaled_data, k))
+Then we use `rowwise` \index{rowwise} + `mutate` to apply the `kmeans` function 
+within each row, to each K.
```
Comment (trevorcampbell) [inline clustering.Rmd:666]: comma misuse

File: clustering.Rmd [hunk 26]
```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
...
-An anonymous function is a function that has no name,
-unlike other functions you've seen so far (`kmeans`, `select`, etc). 
-To do this we will write our map statement like this:
-
-```{r, eval = FALSE}
-map(penguin_clust_ks, function(k) kmeans(scaled_data, k))
+Then we use `rowwise` \index{rowwise} + `mutate` to apply the `kmeans` function 
+within each row, to each K. 
+However, given that the `kmeans` function 
+returns a model object to us, (not a vector),
```
Comment (trevorcampbell) [inline clustering.Rmd:668]: comma misuse

File: clustering.Rmd [hunk 27]
```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
...
+we will need to store the results as a list columm.
+This can work because lists are super vectors, and so they are legitimate
```
Comment (trevorcampbell) [inline clustering.Rmd:670]: "super vectors" -- I know you did some work on wrangling. Do we use the term "super vector" there? I've never heard this term before...

File: clustering.Rmd [hunk 28]
```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
...
 ```
 
-Next, we use `map` again to apply `glance` \index{glance} to each of the K-means
-clustering objects to get the clustering statistics (including WSSD). The output
-of glance is a data frame, and so we get another list column. This results in a 
-complex data frame with 3 columns, one for K, one for the 
+> If we wanted to get one of the clusterings out 
+> of the list column in the data frame,
+> we can use our familiar friends `slice` and `pull`.
+> However, before we do this, we need to turn `rowwise`.
```
Comment (trevorcampbell) [inline clustering.Rmd:693]: "we need to turn rowwise" -- not sure what this means

File: clustering.Rmd [hunk 29]
```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
...
+> To do that, we use `ungroup` 
+> (predictably, this can also turn `group_by` off as well).
```
Comment (trevorcampbell) [inline clustering.Rmd:695]: ah, I guess it should be "turn rowwise off" 

I feel like there must be a better way to say these things than "turn off". Disable? Deactivate?

File: clustering.Rmd [hunk 30]
```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
...
+> penguin_clust_ks |>
+>   ungroup() |>
+>   slice(1) |>
+>   pull(penguin_clusts)
+> ```
+
+Next, we use `mutate` again to apply `glance` \index{glance} 
+to each of the K-means clustering objects to get the clustering statistics 
+(including WSSD). 
+The output of glance is a data frame,
```
Comment (trevorcampbell) [inline clustering.Rmd:707]: glance should be in `code tags`

File: clustering.Rmd [hunk 31]
```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
...
+  rowwise() |>
+  mutate(penguin_clusts = list(kmeans(standardized_data, k)),
+         glanced = list(glance(penguin_clusts)))
+
 penguin_clust_ks
 ```
 
-Finally we extract the total WSSD from the `glanced` column. Given that each 
-item in this column is a data frame, we will need to use the `unnest` function 
+Finally we extract the total WSSD from the `glanced` column.
```
Comment (trevorcampbell) [inline clustering.Rmd:721]: `glanced` is not a function, right? should rephrase this just to use the standard `glance` function name

File: clustering.Rmd [hunk 32]
```diff
@@ -616,18 +745,25 @@ elbow_plot
...
+Unfortunately, for K = 8 we had an unlucky initialization
+and found a bad clustering! \index{K-means!restart,nstart} 
+We can help prevent finding a bad clustering 
+by trying a few different random initializations 
+via the `nstart` argument (here we use 10 restarts). 
+When we do this, K-means clustering will be performed 
+the number of times specified by the `nstart` argument,
+and R will return to us the best clustering from this.
+The more times we perform K-means clustering, 
+the more likely we are to find a lucky, or good clustering (if one exists!).
```
Comment (trevorcampbell) [inline clustering.Rmd:760]: replace with: "the more likely we are to find a good clustering (if one exists)."

File: clustering.Rmd [hunk 33]
```diff
@@ -641,12 +777,13 @@ elbow_plot <- ggplot(clustering_statistics, aes(x = k, y = tot.withinss)) +
 elbow_plot
 ```
 
-
+What value should you choose for `nstart`? The answer is it depends.
```
Comment (trevorcampbell) [inline clustering.Rmd:780]: This paragraph should just be a couple sentences at the end of the last paragraph

**Review by trevorcampbell — COMMENTED** [review]
Overall good! Just lots of writing glitches to fix.

File: clustering.Rmd (hunk 1 above)
Comment (ttimbers) [reply clustering.Rmd:56]: done

File: clustering.Rmd (hunk 2 above)
Comment (ttimbers) [reply clustering.Rmd:70]: Done, changed second to "Given that"

File: clustering.Rmd (hunk 3 above)
Comment (ttimbers) [reply clustering.Rmd:82]: done

File: clustering.Rmd (hunk 5 above)
Comment (ttimbers) [reply clustering.Rmd:101]: I know... but that is what the source says... The station is literally listed as an author here: https://allisonhorst.github.io/palmerpenguins/#references

File: clustering.Rmd (hunk 29 above)
Comment (ttimbers) [reply clustering.Rmd:695]: Actually, we don't need to turn `rowwise` off here (another bug caused me to think this was that, but after fixing that, I just tested it and we don't need to do it. So removing it here.

File: clustering.Rmd (hunk 30 above)
Comment (ttimbers) [reply clustering.Rmd:707]: done

File: clustering.Rmd (hunk 31 above)
Comment (ttimbers) [reply clustering.Rmd:721]: No, the column name is `glanced`. I will make that more clear.

File: clustering.Rmd (hunk 32 above)
Comment (ttimbers) [reply clustering.Rmd:760]: done

File: clustering.Rmd (hunk 33 above)
Comment (ttimbers) [reply clustering.Rmd:780]: done

File: clustering.Rmd [hunk 34]
```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
...
-  mutate(penguin_clusts = map(k, function(ks) kmeans(scaled_data, ks)))
+```{r}
 penguin_clust_ks
 ```
 
-Next, we use `map` again to apply `glance` \index{glance} to each of the K-means
-clustering objects to get the clustering statistics (including WSSD). The output
-of glance is a data frame, and so we get another list column. This results in a 
-complex data frame with 3 columns, one for K, one for the 
+> If we wanted to get one of the clusterings out
```
[the remark this answers is not in the document]
Comment (ttimbers) [reply clustering.Rmd:690]: Done.

**Comment by ttimbers** [comment]
Thanks for the feedback @trevorcampbell - very helpful! Merging!

File: clustering.Rmd (hunk 6 above)
Comment (ttimbers) [reply clustering.Rmd:110]: Yes, I just left it as a record of where the image came from (that is what the image name was at the source. I will just put a comment above with the source link.

File: clustering.Rmd (hunk 7 above)
Comment (ttimbers) [reply clustering.Rmd:115]: fixed

File: clustering.Rmd (hunk 8 above)
Comment (ttimbers) [reply clustering.Rmd:123]: fixed.

File: clustering.Rmd (hunk 10 above)
Comment (ttimbers) [reply clustering.Rmd:133]: `as_double` is soft deprecated in favour of `vctrs::vec_cast`. I have moved to that for our own code - but we don't teach that in the book... What do we teach... hmm...

Comment (ttimbers) [reply clustering.Rmd:133]: This code the students don't see, so I am not going to change it because `vctrs::vec_cast` only works for converting a vector to another vector. What `scale` returns is weird - a matrix. Tidyverse doesn't handle them well, or I don't know the tidyverse version of converting a matrix to a vector (a quick Google search did not help). So I am leaving this as `as.double`. 

Apparently `c` will also convert a matrix to a vector, so could switch to that instead? Still a base R function...

File: clustering.Rmd (hunk 11 above)
Comment (ttimbers) [reply clustering.Rmd:155]: fixed

File: clustering.Rmd (hunk 12 above)
Comment (ttimbers) [reply clustering.Rmd:157]: I changed it to just: "Scatter plot of standardized bill length versus standardized flipper length."

File: clustering.Rmd (hunk 13 above)
Comment (ttimbers) [reply clustering.Rmd:166]: fixed

File: clustering.Rmd (hunk 14 above)
Comment (ttimbers) [reply clustering.Rmd:167]: done

File: clustering.Rmd (hunk 15 above)
Comment (ttimbers) [reply clustering.Rmd:174]: done

File: clustering.Rmd (hunk 16 above)
Comment (ttimbers) [reply clustering.Rmd:176]: fixed

File: clustering.Rmd (hunk 17 above)
Comment (ttimbers) [reply clustering.Rmd:187]: fixed similarly to above

File: clustering.Rmd (hunk 19 above)
Comment (ttimbers) [reply clustering.Rmd:266]: done

File: clustering.Rmd (hunk 20 above)
Comment (ttimbers) [reply clustering.Rmd:298]: The reviewers said this was unclear (the uniformly part in particular), which was why I changed it...

How about a "randomly assigning a roughly equal number of observations"

File: clustering.Rmd (hunk 21 above)
Comment (ttimbers) [reply clustering.Rmd:333]: fixed

File: clustering.Rmd (hunk 22 above)
Comment (ttimbers) [reply clustering.Rmd:399]: fixed

File: clustering.Rmd (hunk 23 above)
Comment (ttimbers) [reply clustering.Rmd:405]: fixed

File: clustering.Rmd (hunk 24 above)
Comment (ttimbers) [reply clustering.Rmd:573]: fixed

File: clustering.Rmd (hunk 25 above)
Comment (ttimbers) [reply clustering.Rmd:666]: fixed

File: clustering.Rmd (hunk 26 above)
Comment (ttimbers) [reply clustering.Rmd:668]: fixed.

File: clustering.Rmd (hunk 27 above)
Comment (ttimbers) [reply clustering.Rmd:670]: fixed.

File: clustering.Rmd (hunk 28 above)
Comment (ttimbers) [reply clustering.Rmd]: We actually don't, so this was fixed in the last PR.
